### PR TITLE
feat(recurrence): am letzten tag des monats (#960) und der anker darunter (#978)

### LINKED ISSUE #978: Eine jaehrliche Serie am 29. Februar rutscht dauerhaft auf den 28.
Found while fixing the monthly clamp in #974. The `FREQ=YEARLY` branch has the same shape of bug the monthly one had, and it survived that fix because it lives in a different branch of `nextOccurrence` (`server/services/recurrence.js`).

Measured:

```
2024-02-29 -> 2025-02-28 -> 2026-02-28 -> 2027-02-28 -> 2028-02-28
```

2028 is a leap year, so the 29th should come back. It does not: the next occurrence is computed from the clamped date, so once a series has fallen to the 28th it never climbs back. A birthday on 29 February is the obvious case, and it is exactly the kind of date where being one day off is noticed.

The monthly branch has the same residue and says so in a comment: clamping produces a throwaway result because nothing remembers the intended day. For monthly, `BYMONTHDAY` (#960) would carry it. For yearly, the equivalent is either `BYMONTHDAY=-1`-style rule data or an anchor column on the series - the same decision, one level up.

Worth deciding together with #960 rather than separately, since both are the same missing anchor.

### PR DESCRIPTION
**Pull request by ulsklyc**
Two wishes with a single cause, which is why they are here together.

The intended day of a recurring series was derived from the **previous
occurrence**. A clamp in a short month therefore rewrote it permanently: a series
begun on 31 January stayed on the 28th from February onwards, and a yearly one on
29 February never came back, not even in the next leap year. The clamp is a
throwaway result; using it as the starting point writes it down.

## Two ways out, and both are in here

**The rule can carry the day.** `BYMONTHDAY` is now read across the full RFC
range, and the repetition form produces exactly one value of it: `-1`, "on the
last day of the month". That is the only statement a start date cannot express,
because it means a different day every month. "On the 15th" needs no field - you
create the series on the 15th. And `BYMONTHDAY=31` would mean something different
here than in RFC 5545, which drops the occurrence in short months instead of
clamping it; a field that means one thing in Yuvomi and another in Apple Calendar
is worse than no field.

**The caller can carry it instead.** Whoever knows DTSTART passes it as an anchor.
The calendar expansion, the ICS parser and the series arithmetic all do - they
iterate from the series start anyway, so it cost them one argument.
`nextDueAfterCompletion` cannot: a task series is a chain of separate rows with no
memory of its origin, and there the previous behaviour stands. No migration, no
existing data touched.

```
FREQ=MONTHLY;BYMONTHDAY=-1   01-31 -> 02-28 -> 03-31 -> 04-30 -> 05-31
FREQ=YEARLY  with anchor     2024-02-29 -> 2025-02-28 -> ... -> 2028-02-29
FREQ=MONTHLY with anchor     01-31 -> 02-28 -> 03-31   (the clamp no longer sticks)
FREQ=MONTHLY without anchor  01-31 -> 02-28 -> 03-28   (unchanged, tasks)
```

A foreign `BYMONTHDAY` stays foreign: the form reads only `-1`, so it rebuilds
only that, and "on the 15th" still falls under the verbatim rule from #756 - it
comes back untouched instead of turning into a checkbox nobody ticked.

## The equivalence test went red, and it was right to

`test-countdown.js` checks that the catch-up **jump** yields the same date as
counting step by step. Its reference counted *without* the anchor while the jump
path passes it through - two different series being compared. `startKey` **is**
the series start, so it is the same anchor; with it both agree again.

Counter-checked that the test still catches real jump errors: remove the guard
against jumping across month ends and it goes red immediately.

## One thing found on the way

`test-rrule-ui.js` held a **hand-copied duplicate** of the server's RRULE regex.
It would have gone stale the moment the validator was extended - and from then on
it would have been testing itself, without ever failing. The test imports the
expression from its source now, and `validate.js` exports it.

## Verification

`npm test` green (4919 assertions, counted by failure lines). Every new assertion
has a counter-proof; the calendar integration got its own, because removing the
anchor there originally broke **nothing** - the unit tests covered
`nextOccurrence` directly and the expansion passing it through was untested.

In the browser: the choice appears only under "monthly", produces
`FREQ=MONTHLY;BYMONTHDAY=-1`, the server accepts it and rejects `BYMONTHDAY=99`.
End to end, a series due 30 September spawns its follow-up on 31 October.

Closes #978

### CODE REVIEW

**Comment by ulsklyc** [comment]
@codex review

The riskiest part is the anchor, because it changes what an existing series computes - so please check it from the callers rather than from the function:

1. `expandRecurringEvents` and the ICS parser now pass DTSTART through. Does that hold for series where DTSTART is *not* itself an occurrence (a `BYDAY` pattern whose start falls outside it, #549), and for the EXDATE skip path?
2. `nextOccurrenceAfter` passes `seriesStart`, which also drives COUNT enforcement. Are those two uses of the same value consistent, or does the anchor change which occurrence is counted as the last one?
3. Task series deliberately get no anchor. Is there a path where a task and a calendar event share a rule and would now disagree - CalDAV round-trip in particular?

Also worth a look: I changed the equivalence test in `test-countdown.js` so its reference counts *with* the anchor. That is the kind of edit that makes a test agree with the code rather than check it, so please judge whether the reference is still an independent one.

File: server/services/recurrence.js [hunk 1]
```diff
@@ -118,20 +156,28 @@ function nextOccurrence(baseDateStr, rrule) {
...
+    const year    = base.getUTCFullYear();
+    const month   = base.getUTCMonth() + interval;
+    const lastDay = new Date(Date.UTC(year, month + 1, 0)).getUTCDate();
+    // `BYMONTHDAY=-1` heisst "letzter Tag", und zwar der des ZIELMONATS - das
+    // ist die eine Angabe, die sich nicht als fester Tag schreiben laesst und
+    // deshalb ueberhaupt in die Regel gehoert (#960).
+    const targetDay = parsed.bymonthday === -1
+      ? lastDay
+      : (parsed.bymonthday ?? anchorDay ?? base.getUTCDate());
+    next.setTime(Date.UTC(year, month, Math.min(Math.max(targetDay, 1), lastDay)));
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/recurrence.js]: Fixed in ad5aabac: negative values now count back from the end of the month as RFC 5545 means them, so -2 is the second-to-last day. You were right that this was worse than before - BYMONTHDAY used to be ignored entirely, so a foreign series at least kept its DTSTART day. A value that parser and validator accept must not be computed wrongly in silence.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/services/recurrence.js [hunk 2]
```diff
@@ -118,20 +156,28 @@ function nextOccurrence(baseDateStr, rrule) {
...
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/recurrence.js]: Fixed in ad5aabac, taking the second option: negatives resolve as lastDay + 1 + bymonthday. Restricting the accepted range would have left foreign rules silently reinterpreted, and the arithmetic is three lines. Both branches share one helper now, so MONTHLY and YEARLY cannot drift apart again.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/services/recurrence.js [hunk 3]
```diff
@@ -118,20 +156,28 @@ function nextOccurrence(baseDateStr, rrule) {
...
-    const targetMonth = base.getUTCMonth();
-    const targetDay   = base.getUTCDate();
-    next.setUTCFullYear(next.getUTCFullYear() + interval);
-    // Feb 29 in non-leap year → Feb 28
-    next.setUTCMonth(targetMonth);
-    const lastDay = new Date(Date.UTC(next.getUTCFullYear(), targetMonth + 1, 0)).getUTCDate();
-    next.setUTCDate(Math.min(targetDay, lastDay));
+    // Monat UND Tag kommen aus dem Anker, wenn es einen gibt: sonst ist der 29.
+    // Februar nach dem ersten Nicht-Schaltjahr fuer immer der 28. (#978).
+    const targetMonth = anchorDate ? anchorDate.getUTCMonth() : base.getUTCMonth();
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/recurrence.js]: Fixed in ad5aabac exactly as suggested - the check is anchorDay !== null now. You are also right about the coverage: the fallback test only exercised MONTHLY, so the YEARLY path had none. It has one now, for both a malformed and an empty anchor.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/services/recurrence.js [hunk 4]
```diff
@@ -255,11 +301,11 @@ function nextOccurrenceAfter(baseDateStr, rrule, notBeforeStr, { seriesStart = n
   if (lastAllowed && notBeforeStr && lastAllowed < notBeforeStr) return null;
 
   const start = notBeforeStr ? fastForward(baseDateStr, parsed, notBeforeStr) : baseDateStr;
-  let current = nextOccurrence(start, rrule);
+  let current = nextOccurrence(start, rrule, { anchor: seriesStart });
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/recurrence.js:322]: Fixed in ad5aabac. A BYMONTHDAY series is not on the grid of its start date, so the cutoff cannot be reached by adding intervals - it is the same situation lastOccurrenceOf already handles for a start day past the 28th, only here the cause is the rule rather than the date. Such series are no longer bounded: one occurrence too many is visible, one too few disappears without a sound. Covered by a test for exactly your example.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: public/rrule-ui.js [hunk 5]
```diff
@@ -71,14 +77,17 @@ export function parseRRule(rule) {
...
   if (!freq) return null;
 
   const parts = [`FREQ=${freq}`];
   if (interval > 1) parts.push(`INTERVAL=${interval}`);
   if (freq === 'WEEKLY' && byday.length > 0) {
     parts.push(`BYDAY=${byday.join(',')}`);
   }
+  // Nur monatlich: "am letzten Tag des Jahres" waere jeder 31. Dezember und
+  // damit ein fester Tag - die Angabe traegt dort nichts bei (#960).
+  if (freq === 'MONTHLY' && lastDay) parts.push('BYMONTHDAY=-1');
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply public/rrule-ui.js:90]: Fixed in ad5aabac. Graph has no absolute last-day, so the mapper emits relativeMonthly with index: last across all seven weekdays - the last occurrence of any weekday is the last day of the month. What cannot be expressed there at all (a negative other than -1, YEARLY with -1) is now not pushed as a recurrence at all rather than as a wrong one. This was the divergence I wanted to avoid by not offering positive BYMONTHDAY values, so thank you for catching that I had reintroduced it through the back door.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: public/rrule-ui.js [hunk 6]
```diff
@@ -61,6 +61,12 @@ export function parseRRule(rule) {
       const n = parseInt(val, 10);
       if (Number.isInteger(n) && n > 0) result.count = n;
     }
+    // NUR `-1`. Jeder andere BYMONTHDAY bleibt fuer dieses Formular eine fremde
+    // Angabe: er wird nicht gelesen, also auch nicht neu gebaut, und faellt damit
+    // unter die Wortlaut-Regel unten - "am 15." kommt aus einem Fremdkalender
+    // unveraendert zurueck, statt hier zu einem Haken zu werden, den niemand
+    // gesetzt hat.
+    if (key === 'BYMONTHDAY' && val.trim() === '-1') result.lastDay = true;
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply public/rrule-ui.js:69]: Fixed in ad5aabac: the summary now carries the qualifier in the bracket, where WEEKLY puts its weekdays - both answer the same question. A test asserts the two series no longer read alike.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: public/rrule-ui.js [hunk 7]
```diff
@@ -194,6 +203,20 @@ export function renderRRuleFields(prefix, existingRule, opts = {}) {
...
+             "Am 15." braucht kein Feld - dafuer legt man die Serie am 15. an.
+             "Am letzten Tag" dagegen meint in jedem Monat einen anderen und ist
+             aus einem Startdatum allein nicht ableitbar: eine am 31. Januar
+             begonnene Serie sah aus wie "letzter Tag" und verlor das im ersten
+             kurzen Monat. Nur bei MONTHLY sichtbar, aus demselben Grund, aus
+             dem buildRRule sie nur dort schreibt. -->
+        <div class="rrule-monthday" id="${prefix}-rrule-monthday" ${parsed.freq === 'MONTHLY' ? '' : 'hidden'}>
+          <label class="toggle" style="margin:0">
+            <input type="checkbox" id="${prefix}-rrule-last-day" ${parsed.lastDay ? 'checked' : ''}>
+            <span>${t('rrule.lastDayOfMonth')}</span>
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply public/rrule-ui.js:217]: Fixed in ad5aabac: the toggle__track span was missing, so the control had neither visible state nor a focus ring - the neighbouring toggle in the same file has it. Checked in the browser this time rather than only in the markup, which is how I missed it: 44x26 track, accent fill, knob translated when checked.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by ulsklyc** [comment]
@codex review

All seven findings are fixed in ad5aabac. Two of them deserve a second look because the fix itself makes a judgement call:

1. **Negative BYMONTHDAY** now resolves as `lastDay + 1 + bymonthday` in a shared helper used by both the MONTHLY and YEARLY branches. Please check the edges: `-31` in February, `31` in a 30-day month, and whether clamping (rather than dropping the occurrence, as RFC 5545 would) is consistent with what the rest of this function does.

2. **The Outlook mapper** emits `relativeMonthly` with `index: "last"` over all seven weekdays for `BYMONTHDAY=-1`. Is that a faithful round-trip - does Graph give back something the ICS/CalDAV side would read as the same rule, or does the series come home as a weekday pattern?

Also: `lastOccurrenceOf` now returns null for any BYMONTHDAY rule, so COUNT stops bounding those series. That trades a possible occurrence too many for never cutting one off, which follows the rule already documented there for start days past the 28th - but it does mean COUNT is no longer enforced for the series this PR introduces. Worth a second opinion on whether that is the right trade.

File: server/services/recurrence.js [hunk 8]
```diff
@@ -35,10 +35,44 @@ function parseRRule(rule) {
...
+ * Die Oberflaeche erzeugt weiterhin nur `-1` (#960); gerechnet wird der ganze
+ * Bereich, weil Fremdkalender ihn liefern.
+ */
+function monthDayFor(bymonthday, lastDay, fallbackDay) {
+  const wanted = bymonthday === null || bymonthday === undefined
+    ? fallbackDay
+    : (bymonthday < 0 ? lastDay + 1 + bymonthday : bymonthday);
+  // Ein `-31` im Februar zeigt vor den Monatsanfang, ein `31` hinter sein Ende:
+  // beide landen am naechstgelegenen echten Tag, wie es die Klemmung immer tat.
+  return Math.min(Math.max(wanted, 1), lastDay);
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/recurrence.js]: Fixed in dd1ff777, by taking the wider range back rather than implementing omission. You are right that clamping an explicit foreign BYMONTHDAY changes the meaning of the data - and once I looked at your other findings together, the pattern was that I had accepted a range the engine does not implement. The parser now reads exactly -1 under FREQ=MONTHLY; every other value is unsupported again, so a foreign series keeps its DTSTART day as it did before this PR.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/services/recurrence.js [hunk 9]
```diff
@@ -35,10 +35,44 @@ function parseRRule(rule) {
...
   const countRaw = parts.COUNT ? parseInt(parts.COUNT, 10) : null;
   const count    = Number.isInteger(countRaw) && countRaw > 0 ? countRaw : null;
+  // BYMONTHDAY traegt den gemeinten Tag IN DER REGEL, statt ihn aus dem letzten
+  // Vorkommen abzuleiten. Die Oberflaeche erzeugt nur `-1` ("letzter Tag des
+  // Monats", #960); gelesen wird der ganze Wertebereich, weil Fremdkalender ihn
+  // liefern und weil ein positiver Wert dieselbe Rolle spielt wie ein Anker.
+  // Mehrere Werte (`BYMONTHDAY=1,15`) sind ausserhalb des Subsets: der erste
+  // gilt, der Rest wird ignoriert - eine Regel, die zwei Tage meint, wuerde von
+  // dieser Funktion sonst still zu einem einzigen.
+  const bmdRaw = parts.BYMONTHDAY ? parseInt(String(parts.BYMONTHDAY).split(',')[0], 10) : null;
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/recurrence.js]: Fixed in dd1ff777: a comma no longer matches, because only the literal "-1" is read. Treating the component as unsupported is exactly what happens now - your point that the previous behaviour (keeping the DTSTART anchor) was better than an arbitrary member is what convinced me to narrow rather than extend.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/services/recurrence.js [hunk 10]
```diff
@@ -118,20 +179,25 @@ function nextOccurrence(baseDateStr, rrule) {
...
+    // `anchorDay !== null`, nicht `anchorDate`: eine Invalid Date ist ein
+    // truthy Objekt, und `getUTCMonth()` daraus ist NaN. Das floss bis in
+    // `Date.UTC` und liess `toISOString()` mit RangeError abbrechen - genau das,
+    // was der Guard oben verhindern soll. Der Fallback-Test deckte nur MONTHLY.
+    const targetMonth = anchorDay !== null ? anchorDate.getUTCMonth() : base.getUTCMonth();
+    const year        = base.getUTCFullYear() + interval;
+    const lastDay     = new Date(Date.UTC(year, targetMonth + 1, 0)).getUTCDate();
+    // Feb 29 in einem Nicht-Schaltjahr → Feb 28, aber ohne den Anker zu verlieren.
+    const targetDay = monthDayFor(parsed.bymonthday, lastDay, anchorDay ?? base.getUTCDate());
+    next.setTime(Date.UTC(year, targetMonth, targetDay));
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/recurrence.js:198]: Fixed in dd1ff777 by leaving BYMONTHDAY unsupported for YEARLY, as you suggested. Implementing the twelve-occurrence expansion would be a different feature; the parser now only reads the component under FREQ=MONTHLY.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/middleware/validate.js [hunk 11]
```diff
@@ -18,7 +18,10 @@ const DATETIME_RE = /^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}(:\d{2}(?:\.\d+)?)?(?:Z|[+-]
 const COLOR_RE    = /^#[0-9A-Fa-f]{6}$/;
 const MONTH_RE    = /^\d{4}-\d{2}$/;
 // UNTIL und COUNT schließen sich laut RFC 5545 gegenseitig aus (#513).
-const RRULE_RE    = /^(FREQ=(DAILY|WEEKLY|MONTHLY|YEARLY)(;INTERVAL=\d{1,2})?(;BYDAY=[A-Z,]{2,}(,[A-Z]{2})*)?(;(UNTIL=\d{8}(T\d{6}Z)?|COUNT=\d{1,4}))?)?$/;
+// BYMONTHDAY steht zwischen BYDAY und der Endbedingung, weil die Oberflaeche
+// die Regel in dieser Reihenfolge baut. Erlaubt ist der volle RFC-Bereich
+// (-31..-1, 1..31); erzeugt wird aus dem Formular nur `-1` (#960).
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/middleware/validate.js]: Fixed in dd1ff777: the regex accepts only BYMONTHDAY=-1 now. Your example was the clearest argument for narrowing everything else - a rule that saves successfully and then does something other than what it says is worse than one that is rejected. Covered by a test that asserts the validator rejects 15, 31, -2, -31, 0 and 1,15.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/services/outlook-calendar.js [hunk 12]
```diff
@@ -470,18 +470,36 @@ function rruleToGraphRecurrence(rrule, startDate, tz = outlookTimeZone()) {
...
+    // `absoluteMonthly` nimmt nur eine feste Zahl, und die waere hier der Tag
+    // des Startdatums: eine am 15. begonnene Serie liefe in Yuvomi, CalDAV,
+    // Google und ICS auf dem Monatsletzten, in Outlook aber auf dem 15. - eine
+    // Divergenz, die niemand bemerkt, bis die Termine auseinanderlaufen.
+    //
+    // `relativeMonthly` mit `index: 'last'` ueber ALLE sieben Wochentage sagt
+    // dasselbe: das letzte Vorkommen irgendeines Wochentags im Monat ist dessen
+    // letzter Tag. Das ist der uebliche Weg, die Angabe in Graph auszudruecken.
+    pattern = parsed.bymonthday === -1
+      ? { type: 'relativeMonthly', interval: parsed.interval, daysOfWeek: [...GRAPH_DAYS], index: 'last' }
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/outlook-calendar.js]: Fixed in dd1ff777 - you were right and my substitute was wrong. The recurrence is now omitted entirely for such a series rather than sent as a different one. I had not checked how Graph resolves multiple daysOfWeek, and the ICS round-trip argument settles it independently.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/services/outlook-calendar.js [hunk 13]
```diff
@@ -470,18 +470,36 @@ function rruleToGraphRecurrence(rrule, startDate, tz = outlookTimeZone()) {
...
   } else {
     return null;
   }
 
+  // LIEBER GAR KEINE WIEDERHOLUNG ALS EINE FALSCHE. Ein negativer BYMONTHDAY
+  // ausser `-1` zaehlt vom Monatsende rueckwaerts ("vorletzter Tag"), und dafuer
+  // hat Graph keine Entsprechung. Eine solche Regel kommt nur aus einem
+  // Fremdkalender; sie als absoluten Tag zu pushen verschoebe jeden Termin.
+  if (parsed.bymonthday !== null && parsed.bymonthday < 0 && parsed.bymonthday !== -1) return null;
+  if (parsed.freq === 'YEARLY' && parsed.bymonthday === -1) return null;
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/outlook-calendar.js]: Fixed in dd1ff777: the payload now sets recurrence: null explicitly instead of omitting the field, so PATCH clears the stale recurrence rather than leaving it. The detail that the content hash is saved as if the update had converged is what makes this one nasty - it would have stayed wrong silently.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/services/recurrence.js [hunk 14]
```diff
@@ -277,9 +343,17 @@ function nextOccurrenceAfter(baseDateStr, rrule, notBeforeStr, { seriesStart = n
...
+  const { freq, interval, byday, count, bymonthday } = parsed;
   if (!count || byday.length) return null;
 
+  /* EINE REGEL MIT BYMONTHDAY LAEUFT NICHT AUF DEM RASTER IHRES STARTDATUMS.
+   * `FREQ=MONTHLY;BYMONTHDAY=-1;COUNT=3` ab dem 15. Januar bedeutet Jan 15,
+   * Feb 28, Mrz 31 - (COUNT - 1) Intervalle ab dem Start ergeben dagegen den
+   * 15. Maerz, und eine Grenze dort schneidet das letzte Vorkommen ab. Dieselbe
+   * Ueberlegung wie beim Tag > 28 gleich darunter, nur ist der Grund hier die
+   * Regel und nicht das Datum. */
+  if (bymonthday !== null && bymonthday !== undefined) return null;
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/recurrence.js]: Fixed in dd1ff777. Disabling the bound entirely was the wrong answer to the early-cutoff problem, as your COUNT=1 example shows. For a -1 series the bound is exactly computable, because every occurrence except the first is the last day of its month and DTSTART is occurrence 1. Tested in both directions: the final occurrence survives, and the series ends after it.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: docs/SPEC.md [hunk 15]
```diff
@@ -576,7 +576,7 @@ them with 400 (see the Reminders section), because the run would recreate them w
...
...sks and calendar events alike, since `expandRecurringEvents` walks the same function. **Both remaining halves of that were closed in v2.61.0** (#960, #978), because they had one cause: the intended day was derived from the *previous* occurrence, and a clamp in a short month therefore rewrote it for good. Two ways out, and the code carries both. The **rule** can name the day: `BYMONTHDAY` is read across the full RFC range, and the repetition form produces exactly one value of it - `-1`, "on the last day of the month". That is the one statement a start date cannot express, since it means a different day every month; "on the 15th" needs no field, because you create the series on the 15th. The **caller** can name it instead: whoever knows DTSTART passes it as an anchor, which the calendar expansion, the ICS parser and the series arithmetic all do - they iterate from the series start anyway. `nextDueAfterCompletion` cannot, because a task series is a chain of separate rows that does not know its origin, and there the previous behaviour stands. Any BYMONTHDAY other than `-1` stays a foreign rule that the form does not read and therefore does not rebuild, so it survives an edit verbatim |
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply docs/SPEC.md]: Fixed in dd1ff777: toLocalRRule now passes BYMONTHDAY=-1 through (and only that). While there I found the same hand-copied validator regex in that suite that I had just removed from test-rrule-ui.js - it imports the real one now, so the parity test cannot drift from what it claims to check.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by ulsklyc** [comment]
@codex review

All eight fixed in dd1ff777 - and the fix is a **retreat**, not an extension. Your findings shared one cause: I accepted `BYMONTHDAY` across the full RFC range without implementing its semantics, so each opening created a failure case. It is now read only as the literal `-1`, only under `FREQ=MONTHLY`, in the parser, the validator regex and the ICS import reducer alike. Everything else is unsupported again and a foreign series keeps its DTSTART day, exactly as before this PR.

Two things worth a fresh look, because they are new rather than narrowed:

1. `lastOccurrenceOf` now computes the COUNT bound for a `-1` series directly (DTSTART is occurrence 1; every later one is the last day of its month). Please check the arithmetic against `INTERVAL > 1` and a series whose DTSTART already falls on a month end.
2. `localEventToGraph` sets `recurrence: null` explicitly when the rule cannot be mapped. Does Graph accept that on PATCH as "make this a single occurrence", and does it do anything unwanted on the initial POST, where nothing needs clearing?

File: server/middleware/validate.js [hunk 16]
```diff
@@ -18,7 +18,14 @@ const DATETIME_RE = /^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}(:\d{2}(?:\.\d+)?)?(?:Z|[+-]
...
 // UNTIL und COUNT schließen sich laut RFC 5545 gegenseitig aus (#513).
-const RRULE_RE    = /^(FREQ=(DAILY|WEEKLY|MONTHLY|YEARLY)(;INTERVAL=\d{1,2})?(;BYDAY=[A-Z,]{2,}(,[A-Z]{2})*)?(;(UNTIL=\d{8}(T\d{6}Z)?|COUNT=\d{1,4}))?)?$/;
+// BYMONTHDAY steht zwischen BYDAY und der Endbedingung, weil die Oberflaeche
+// die Regel in dieser Reihenfolge baut. Erlaubt ist AUSSCHLIESSLICH `-1`
+// ("letzter Tag des Monats", #960) - die erste Fassung nahm den vollen
+// RFC-Bereich an, obwohl die Engine ihn nicht bedient: `FREQ=WEEKLY;
+// BYMONTHDAY=15` liess sich speichern und lief danach woechentlich, ohne den
+// angenommenen Monatstag je anzuwenden. Was nicht implementiert ist, wird nicht
+// angenommen.
+const RRULE_RE    = /^(FREQ=(DAILY|WEEKLY|MONTHLY|YEARLY)(;INTERVAL=\d{1,2})?(;BYDAY=[A-Z,]{2,}(,[A-Z]{2})*)?(;BYMONTHDAY=-1)?(;(UNTIL=\d{8}(T\d{6}Z)?|COUNT=\d{1,4}))?)?$/;
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/middleware/validate.js]: Fixed: the regex now has two branches - MONTHLY may carry the component, the other frequencies may not. You are right that this was the same "accepted but never honored" failure one level up. My test for it was blind: it only exercised MONTHLY rules, so it could not see the case at all. It now runs across all four frequencies with a counter-check per frequency.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/middleware/validate.js]: Fixed, and thank you for spelling out that the group was a sibling of the alternation rather than nested - that is exactly what my own test could not see, because it only ever built MONTHLY rules. Two branches now; FREQ=WEEKLY/DAILY/YEARLY with BYMONTHDAY=-1 are rejected.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: public/rrule-ui.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (ulsklyc) [reply public/rrule-ui.js:90]: Fixed: truncateRuleBefore keeps BYMONTHDAY=-1 (only that value, only under MONTHLY, and the result is checked against the real validator regex). Rebuilding a rule from its parts and dropping what you do not know is the same silent rewrite that #756 guards against elsewhere.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: server/services/ics-subscription.js [hunk 17]
```diff
@@ -309,6 +309,12 @@ function toLocalRRule(raw) {
       .filter((d) => /^(MO|TU|WE|TH|FR|SA|SU)$/.test(d));
     if (days.length) rule += `;BYDAY=${days.join(',')}`;
   }
+  // Genau `-1` bei MONTHLY, in der Reihenfolge, die der Validator erwartet
+  // (#960). Der Reduzierer laesst alles weg, was die Engine nicht bedient - und
+  // seit "am letzten Tag des Monats" darstellbar ist, gehoert es dazu. Ohne
+  // diese Zeile kaeme eine importierte Monatsletzten-Serie als blosses
+  // `FREQ=MONTHLY` an und liefe danach auf dem Tag ihres Startdatums.
+  if (freq === 'MONTHLY' && String(parts.BYMONTHDAY ?? '').trim() === '-1') rule += ';BYMONTHDAY=-1';
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/ics-subscription.js:317]: Fixed by making COUNT advance only for dates that satisfy BYDAY, in both expansions. Worth noting the bug is older than this PR and reproduces without BYMONTHDAY at all: FREQ=MONTHLY;BYDAY=MO;COUNT=2 returned a single occurrence, because the second count went to a Wednesday nobody ever saw. The BYDAY filter and the EXDATE check sat in one condition after the increment - they look alike and are not: a day outside the pattern is not an occurrence, an excluded one is and must still count (RFC 5545, #513). Tested in both directions so splitting them cannot tip the other half.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by ulsklyc** [comment]
@codex review

All four fixed in a2f7bfa2. One of them is worth flagging because it reaches beyond this feature:

**The COUNT/BYDAY miscount is older than this PR and reproduces without BYMONTHDAY at all.** `FREQ=MONTHLY;BYDAY=MO;COUNT=2` returned a single occurrence, because the increment happened before the BYDAY filter and the second count went to a Wednesday that was never emitted. I split the two conditions in both expansions - a day outside the pattern is not an occurrence and must not count, an EXDATE-excluded one is and must (RFC 5545, #513). Please check that split from the other side: is there a caller that relied on the old counting, and does `expandRRULE` in the ICS parser now agree with `expandRecurringEvents` for the same rule?

The validator regex was rebuilt as two branches (MONTHLY may carry BYMONTHDAY, the others may not). It is assembled from string fragments now rather than written as one literal - please confirm the two branches accept exactly the same rules as before apart from the intended restriction.

File: server/services/calendar-events.js [hunk 18]
```diff
@@ -94,13 +98,30 @@ export function expandRecurringEvents(events, from, to, exceptionsByEvent = null
...
+      // Beide standen bis hierher in EINER Bedingung nach `occurrence++`, und
+      // damit verbrauchte jeder uebersprungene Wochentag ein Vorkommen:
+      // `FREQ=MONTHLY;BYDAY=MO;COUNT=2` lieferte genau einen Termin, weil der
+      // zweite Zaehler an einen Mittwoch ging, den niemand je zu sehen bekam.
+      if (!matchesRRuleByday(currentDate, event.recurrence_rule)) {
+        const next = nextOccurrence(currentDate, event.recurrence_rule, { anchor: seriesStart });
+        if (!next || next <= currentDate) break;
+        currentDate = next;
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (ulsklyc) [reply server/services/calendar-events.js:118]: You are right on both counts, and I reproduced it before changing anything: the series expanded to 2026-01-15, 02-28, 03-31 instead of 01-31, 02-28, 03-31 - one appointment on a day that is not an occurrence, and one that never existed.

Fixed in https://github.com/ulsklyc/yuvomi/pull/984, since this PR was already merged. `matchesRRuleByday` now checks BYMONTHDAY as well (its name says one thing and its job was always the other), and the next occurrence may fall in the same month when the month end is still ahead. Two of my own tests held the wrong result and went red for it.

The export half stands as you describe it and is called out in that PR rather than fixed: the stored DTSTART still goes out unsynchronised, and the only real remedy is to move the series start when saving - which changes what the user typed and is a separate decision.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
